app/kubemci/pkg/ingress/ingresssyncer.go

```diff
@@ -0,0 +1,156 @@
+package ingress
+
+import (
+	"fmt"
+	"io/ioutil"
+	"reflect"
+
+	"github.com/ghodss/yaml"
+	"github.com/golang/glog"
+	multierror "github.com/hashicorp/go-multierror"
+	"k8s.io/api/extensions/v1beta1"
+	"k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/util/diff"
+	kubeclient "k8s.io/client-go/kubernetes"
+	"k8s.io/ingress-gce/pkg/annotations"
+)
+
+var (
+	defaultIngressNamespace = "default"
+)
+
+type IngressSyncer struct {
+}
+
+func NewIngressSyncer() IngressSyncerInterface {
+	return &IngressSyncer{}
+}
+
+// EnsureIngress ensures that the ingress exists in all clusters given.
+// Does nothing if it already exists in a cluster, else creates the resource.
+// Returns a list of clusters where the ingress resource has been ensured.
+func (i *IngressSyncer) EnsureIngress(ing *v1beta1.Ingress, clients map[string]kubeclient.Interface, forceUpdate bool) ([]string, error) {
+	var err error
+	var ensuredClusters []string
+	for cluster, client := range clients {
+		glog.V(4).Infof("Creating Ingress in cluster: %v...", cluster)
+		ensuredClusters = append(ensuredClusters, cluster)
+		glog.V(3).Infof("Using namespace %s for ingress %s", ing.Namespace, ing.Name)
+		existingIng, getErr := client.ExtensionsV1beta1().Ingresses(ing.Namespace).Get(ing.Name, metav1.GetOptions{})
+		if getErr != nil {
+			if errors.IsNotFound(getErr) {
+				actualIng, createErr := client.ExtensionsV1beta1().Ingresses(ing.Namespace).Create(ing)
+				glog.V(2).Infof("Ingress Create returned: err:%v. Actual Ingress:%+v\n", createErr, actualIng)
+				if createErr != nil {
+					err = multierror.Append(err, fmt.Errorf("Error in creating ingress in cluster %s: %s", cluster, createErr))
+				}
+			} else {
+				err = multierror.Append(err, fmt.Errorf("Error in checking for existing ingress in cluster %s: %s", cluster, getErr))
+			}
+			fmt.Println("Created Ingress in cluster:", cluster)
+		} else {
+			if !reflect.DeepEqual(ing, existingIng) {
```

Comment:
We will need to be smarter here.
existingIng will have extra fields like CreateTimestamp, uid, maybe the status will be set. reflect.DeepEqual will never succeed.

Reply:
ObjectMetaAndSpecEquivalent as defined here: https://github.com/kubernetes/federation/blob/7951a643cebc3abdcd903eaff90d1383b43928d1/pkg/federation-controller/util/meta.go#L79 seems like what we need.